includes/wcsr-time-functions.php

```diff
@@ -0,0 +1,134 @@
+		$next_timestamp         = wcs_add_time( $billing_interval, $billing_period, $from_timestamp ); // use the same function used to calculate subscription billing periods to work out the exact billing
+		$days_in_billing_period = wcsr_get_days_in_period( $from_timestamp, $next_timestamp );
+		$days_in_billing_cycle += $days_in_billing_period;
+
+		if ( $days_left_in_period >= $days_in_billing_period ) {
+			$number_of_billing_periods++;
+		} elseif ( $days_left_in_period < $days_in_billing_period ) { // if the days left in the period are less then the days in the billing cycle - find the ratio of days left and days in billing period
+			$number_of_billing_periods += round( $days_left_in_period / $days_in_billing_period, 2 );
+			break;
+		}
+
+		$counter++;
+		$days_left_in_period -= $days_in_billing_period;
+		$from_timestamp       = $next_timestamp;
+	}
+
+	// if the number of days active is more than or equal to the days in the period - return the exact number of billing periods (remember this can be a float or whole number i.e. 15 day period of a monthly subscription is only 0.5)
+	if ( $days_active >= $days_in_period ) {
+		$active_days_ratio = $number_of_billing_periods;
+	} else {
+		// calculate the active days ratio for days active and days in period, then multiply that by the number of billing periods found between the from and to timestamp. This replaces the extra days in cycle logic because instead of finding the extra days and minusing that off the days in period, we can just multiply the result by the number of cycles.
+		$active_days_ratio = round( ( $days_active / $days_in_period ) * $number_of_billing_periods, 2 );
+	}
+
+	return $active_days_ratio;
+}
+
+/**
+ * Calculates the number of whole (i.e. using floor) days between two timestamps.
+ * Uses the same logic in @see wcs_estimate_periods_between() but this function exclusively calculates the number of days
+ *
+ * @since 1.1.0
+ * @param int $start_timestamp The starting timestamp to calculate the number of days from
+ * @param int $end_timestamp   The end timestamp to calculate the number of days to
+ * @return int
+ */
+function wcsr_get_days_in_period( $start_timestamp, $end_timestamp ) {
+	// return 0 if the start timestamp is after the end timestamp
+	$periods_until = 0;
+
+	if ( $end_timestamp > $start_timestamp ) {
+		$periods_until = floor( ( $end_timestamp - $start_timestamp ) / DAY_IN_SECONDS );
+	}
+
+	return $periods_until;
+}
+
+/**
+ * Find all the timestamps from a given array that fall within a from/to timestamp range.
+ *
+ * @param array $timestamps_to_check
+ * @param int $from_timestamp
+ * @param int $to_timestamp
+ * @return array
+ */
+function wcsr_get_timestamps_between( $timestamps_to_check, $from_timestamp, $to_timestamp ) {
+
+	$times = array();
+
+	foreach ( $timestamps_to_check as $i => $timestamp ) {
+		if ( $timestamp >= $from_timestamp && $timestamp <= $to_timestamp ) {
+			$times[ $i ] = $timestamp;
```

Comment:
Following from https://github.com/Prospress/woocommerce-subscriptions-resource/pull/16/files#r175308497 I suspect `$timestamp `should just be popped onto the `$times` array. I've found an instance where the days calculated returned is 31 (well in prod it was 32 😢) but looking at the resource activation/deactivation times should have only been 5 (or 4)

Reply:
I should note adjust this to be `$times[] = $timestamp;` corrects the calcs in `get_days_active()`